### PR TITLE
fix: too many io requests for read blocks during compact

### LINKED ISSUE #9107: bug: 10,000 level block compact
### Search before asking

- [X] I had searched in the [issues](https://github.com/datafuselabs/databend/issues) and found no similar issues.

### Version

v0.8.142

### What's Wrong?

databend on local minio 
```
  MySQL [(none)]> select count(*)  from fuse_block('default','hits');
+----------+
| count(*) |
+----------+
|    13074 |
+----------+
1 row in set (1.003 sec)
optimize table hits compact;  

return error:
ERROR 1105 (HY000): Code: 4000, displayText = Unexpected (persistent) at read, context: { called: http_util::Client::send_async, service: s3, path: 1/94363/_b/945a8f3d295641ce81642616173d6051_v0.parquet, range: 247122-247208 } => send async request, source: error sending request for url (http://127.0.0.1:9900/testbucket/local_7150/1/94363/_b/945a8f3d295641ce81642616173d6051_v0.parquet): error trying to connect: tcp connect error: Address not available (os error 99).

netstat -an | grep 9900  |awk '{print $6}' |sort |uniq -c |sort -nr
 110485 ESTABLISHED
     40 TIME_WAIT
      3 FIN_WAIT1
      1 LISTEN
      1 FIN_WAIT2
```

### How to Reproduce?

_No response_

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by zhyass**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

This bug is due to the fact that compact transform reads and writes blocks concurrently, and there was no limit to the number of concurrent requests before, which may lead to excessive io requests and tcp connect error.
```
MySQL [(none)]> optimize table hits compact;
ERROR 1105 (HY000): Code: 4000, displayText = Unexpected (persistent) at read, context: { called: http_util::Client::send_async, service: s3, path: 1/94363/_b/02c89d8096504e4c97fde3e56sd17875_v0.parquet, range: 426216-426309 } => send async request, source: error sending request for url (http://127.0.0.1:9900/testbucket/local_7150/1/94363/_b/02c89d8096504e4c97fde3e56sd17875_v0.parquet): error trying to connect: tcp connect error: Address not available (os error 99).
```

So we need limit the number of concurrency(max_io_requests) when [reading](https://github.com/datafuselabs/databend/blob/main/src/query/storages/fuse/fuse/src/operations/mutation/compact/compact_transform.rs#L344-L352) and [writing](https://github.com/datafuselabs/databend/blob/main/src/query/storages/fuse/fuse/src/operations/mutation/compact/compact_transform.rs#L362-L370) blocks.

https://github.com/datafuselabs/databend/blob/main/src/query/storages/fuse/fuse/src/operations/mutation/compact/compact_transform.rs#L344-L352

We can avoid oom and improve the efficiency of compact by adjusting `max_threads`| `max_storage_io_requests`| `max_memory_usage`.

Closes #9107

### CODE REVIEW

**Comment by BohuTANG** [comment]
@wubx we can do a compact on your table with PR, and check the performance.

**Comment by wubx** [comment]
LGTM

File: src/query/storages/fuse/fuse/src/io/segments.rs [hunk 1]
```diff
@@ -74,22 +71,36 @@ impl SegmentsIO {
...
-        // 1.4 get all the result.
-        let joint = future::try_join_all(join_handlers)
-            .instrument(tracing::debug_span!("read_segments_join_all"))
-            .await
-            .map_err(|e| ErrorCode::StorageOther(format!("read segments failure, {}", e)))?;
+    // 1. build the runtime.
+    let semaphore = Semaphore::new(max_io_requests);
+    let segments_runtime = Arc::new(Runtime::with_worker_threads(
+        max_runtime_threads,
+        Some("deletion-write-segments-worker".to_owned()),
```
Comment (dantengsky) [inline src/query/storages/fuse/fuse/src/io/segments.rs:93]: maybe better let caller site decides the thread name？

Comment (zhyass) [reply src/query/storages/fuse/fuse/src/io/segments.rs]: > maybe better let caller site decides the thread name？

Added

File: src/query/storages/fuse/fuse/src/operations/mutation/compact/compact_transform.rs [hunk 2]
```diff
@@ -182,7 +182,7 @@ impl CompactTransform {
             stats_of_columns.push(meta_stats);
         }
 
-        let blocks = try_join_futures(ctx, task_futures)
+        let blocks = try_join_futures(ctx, task_futures, "deletion-read-blocks-worker".to_owned())
```
Comment (dantengsky) [inline src/query/storages/fuse/fuse/src/operations/mutation/compact/compact_transform.rs:185]: :+1:  but why prefixed with "deletion-"?

**Comment by dantengsky** [comment]
@wubx help :) @zhyass has added some extra tweaks to control the peak memory usage, would you please re-check the behavior of compaction is as expected? (the CPU usage and the peak memory usage)

**Comment by wubx** [comment]
> @wubx help :) @zhyass has added some extra tweaks to control the peak memory usage, would you please re-check the behavior of compaction is as expected or not? (the CPU usage and the peak memory usage)

Let me check. 
I will migration full hits from clickhouse to Databend using Addax, Finally, execute compact.

**Comment by wubx** [comment]
LGTM, Greatly!
Show you the results:
```
MySQL [(none)]> select * from fuse_snapshot('default','hits') limit 1\G;
*************************** 1. row ***************************
         snapshot_id: c8f21d5a6a82451281c9978e3bd9726a
   snapshot_location: 1/238690/_ss/c8f21d5a6a82451281c9978e3bd9726a_v1.json
      format_version: 1
previous_snapshot_id: 04ba4122ead841eaa419ce8b52fa619c
       segment_count: 48829
         block_count: 48829
           row_count: 99997497
  bytes_uncompressed: 81381357474
    bytes_compressed: 21929043037
          index_size: 2088644102
           timestamp: 2022-12-08 02:34:28.561389
1 row in set (0.008 sec)

ERROR: No query specified

MySQL [(none)]> optimize table hits compact;
Query OK, 0 rows affected (37 min 8.238 sec)

MySQL [(none)]>
MySQL [(none)]>
MySQL [(none)]> select * from fuse_snapshot('default','hits') limit 2\G;
*************************** 1. row ***************************
         snapshot_id: a7b1f0fb54714d45b46d2f08da3f7f9c
   snapshot_location: 1/238690/_ss/a7b1f0fb54714d45b46d2f08da3f7f9c_v1.json
      format_version: 1
previous_snapshot_id: c8f21d5a6a82451281c9978e3bd9726a
       segment_count: 49
         block_count: 750
           row_count: 99997497
  bytes_uncompressed: 81370587778
    bytes_compressed: 19819963262
          index_size: 699928516
           timestamp: 2022-12-08 03:12:43.226270
*************************** 2. row ***************************
         snapshot_id: c8f21d5a6a82451281c9978e3bd9726a
   snapshot_location: 1/238690/_ss/c8f21d5a6a82451281c9978e3bd9726a_v1.json
      format_version: 1
previous_snapshot_id: 04ba4122ead841eaa419ce8b52fa619c
       segment_count: 48829
         block_count: 48829
           row_count: 99997497
  bytes_uncompressed: 81381357474
    bytes_compressed: 21929043037
          index_size: 2088644102
           timestamp: 2022-12-08 02:34:28.561389
2 rows in set (0.009 sec)
```
image

**Review by BohuTANG — APPROVED** [review]

**Pull request merged**